operator/chartdownloader/factory.go

```diff
@@ -0,0 +1,54 @@
+package chartdownloader
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws/client"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"k8s.io/helm/pkg/proto/hapi/chart"
+)
+
+type ChartDownloader interface {
+	Download(context.Context, string) (*chart.Chart, error)
+}
+
+// TODO: github, GCS support?
+type Providers interface {
```

Comment:
I'm open to ideas for how to do this differently/better. Dependency injection for a factory-thing is awkward, since you have to provide all the dependencies for every possible output option. This interface/struct-of-funcs pattern allows only defining the ones you expect to use, and missing providers are handled gracefully, but I'm sure it could be improved.

Reply:
We can worry about this later.